### PR TITLE
LR support in DomainDecomposition

### PR DESCRIPTION
**Pull request by akva2**
this adds unstructured patch support in parallel. all node numbers are covered by the same tests as we have for tensor splines, however the equation numbers are not as thoroughly tested, only confirmation that we have the expect amount of equations on each process is performed.

Note: This requires an updated LR spline library.

### CODE REVIEW

**Comment by akva2** [comment]
This is now fully tested for single basis. I will append some mixed tests, but it is ready for review.

**Comment by akva2** [comment]
jenkins build this please

**Comment by notspiff** [comment]
jenkins build this please

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
jenkins build this please

**Comment by akva2** [comment]
there. mixed is tested as well (2d only, no mixed 3d lr support in general yet). i now consider this to be complete.

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-advectiondiffusion=14 ifem-stokes=70 ifem-navierstokes=68 ifem-poisson=7 please

File: src/ASM/ASM3D.h [hunk 1]
```diff
@@ -164,7 +164,7 @@ class ASM3D
   //! \brief Returns the node index for a given corner.
   virtual int getCorner(int I, int J, int K, int basis) const = 0;
   //! \brief Returns the node indices for a given edge.
-  virtual std::vector<int> getEdge(int lEdge, bool open, int basis) const = 0;
+  virtual std::vector<int> getEdge(int lEdge, bool open, int basis, int orient=0) const = 0;
```
Comment (kmokstad) [inline src/ASM/ASM3D.h]: space round = in the initializer

File: src/ASM/LR/ASMu2Dmx.h [hunk 2]
```diff
@@ -182,6 +182,15 @@ class ASMu2Dmx : public ASMu2D, private ASMmxBase
   virtual bool connectPatch(int edge, ASM2D& neighbor, int nedge, bool revers,
                             int = 0, bool coordCheck = true, int thick = 1);
 
+  //! \brief Finds the global (or patch-local) node numbers on a patch boundary.
+  //! \param[in] lIndex Local index of the boundary edge
+  //! \param nodes Array of node numbers
+  //! \param basis Which basis to grab nodes for (0 for all)
+  //! \param thick Thickness of connection
+  //! \param local If \e true return patch-local node numbers
+  virtual void getBoundaryNodes(int lIndex, IntVec& nodes, int basis,
```
Comment (kmokstad) [inline src/ASM/LR/ASMu2Dmx.h]: Mark the last three arguments as [in] in the doxy. orient is not documented

File: src/ASM/LR/ASMu3D.h [hunk 3]
```diff
@@ -92,14 +92,15 @@ class ASMu3D : public ASMunstruct, public ASM3D
   //! \param[in] lIndex Local index of the boundary face
   //! \param nodes Array of node numbers
   //! \param basis Which basis to grab nodes for (0 for all)
+  //! \param orient Orientation of boundary
   //! \param local If \e true, return patch-local node numbers
   virtual void getBoundaryNodes(int lIndex, IntVec& nodes,
```
Comment (kmokstad) [inline src/ASM/LR/ASMu3D.h]: Mark the input arguments as [in] in the doxy

File: src/ASM/ASMunstruct.h [hunk 4]
```diff
@@ -133,8 +133,12 @@ class ASMunstruct : public ASMbase
   //! \brief Sort basis functions based on local knot vectors.
   static void Sort(int u, int v, int orient, std::vector<LR::Basisfunction*>& functions);
 
+  //! \brief Returns basis used for geometry.
+  int getGeoBasis() const { return lrGeoBasis; }
+
 protected:
   LR::LRSpline* geo; //!< Pointer to the actual spline geometry object
+  int lrGeoBasis = 1; //!< Basis used for geometry.
```
Comment (kmokstad) [inline src/ASM/ASMunstruct.h]: Prefer to do the initialization in the constructor

File: src/ASM/LR/ASMu2Dmx.C [hunk 5]
```diff
@@ -176,10 +176,34 @@ bool ASMu2Dmx::generateFEMTopology ()
...
+    } else if (lrspline) {
+      if (ASMmxBase::Type == REDUCED_CONT_RAISE_BASIS1 ||
+          ASMmxBase::Type == REDUCED_CONT_RAISE_BASIS2) {
+        m_basis.resize(2);
+        m_basis[0].reset(lrspline->getRaiseOrderSpace(1,1));
+        m_basis[1] = lrspline;
+      } else if (ASMmxBase::Type == FULL_CONT_RAISE_BASIS1 ||
+                 ASMmxBase::Type == FULL_CONT_RAISE_BASIS2) {
+        m_basis.resize(2);
+        m_basis[1].reset(lrspline->getDerivedBasis(-1,-1,1,1));
```
Comment (VikingScientist) [inline src/ASM/LR/ASMu2Dmx.C]: `getDerivedBasis()` har fått en bugfix i [multipatch-branchen](https://github.com/VikingScientist/LRsplines/tree/Multipatch-matching) til LR

File: src/ASM/ASM3D.h (hunk 1 above)
Comment (akva2) [inline src/ASM/ASM3D.h]: done.

File: src/ASM/LR/ASMu2Dmx.h (hunk 2 above)
Comment (akva2) [inline src/ASM/LR/ASMu2Dmx.h]: done and done.

File: src/ASM/LR/ASMu3D.h (hunk 3 above)
Comment (akva2) [inline src/ASM/LR/ASMu3D.h]: done.

File: src/ASM/ASMbase.h [hunk 6]
```diff
@@ -214,10 +214,11 @@ class ASMbase
   //! \param nodes Array of node numbers
   //! \param basis Which basis to grab nodes for (0 for all)
   //! \param thick Thickness of connection
+  //! \param[in] orient Local orientation of the boundary face/edge
   //! \param local If true, return patch-local numbers
   virtual void getBoundaryNodes(int lIndex, IntVec& nodes,
```
Comment (kmokstad) [inline src/ASM/ASMbase.h]: Mark other arguments also as [in] and the `nodes` as [out]

File: src/ASM/LR/ASMu2D.h [hunk 7]
```diff
@@ -96,9 +96,10 @@ class ASMu2D : public ASMunstruct, public ASM2D
   //! \param[in] lIndex Local index of the boundary edge
   //! \param nodes Array of node numbers
   //! \param basis Which basis to grab nodes for (0 for all)
+  //! \param orient Orientation of boundary (used for sorting)
   //! \param local If \e true return patch-local node numbers
   virtual void getBoundaryNodes(int lIndex, IntVec& nodes,
```
Comment (kmokstad) [inline src/ASM/LR/ASMu2D.h]: Mark arguments [in] and `nodes` as [out]

File: src/ASM/LR/ASMu2Dmx.h [hunk 8]
```diff
@@ -182,6 +182,17 @@ class ASMu2Dmx : public ASMu2D, private ASMmxBase
   virtual bool connectPatch(int edge, ASM2D& neighbor, int nedge, bool revers,
                             int = 0, bool coordCheck = true, int thick = 1);
 
+  //! \brief Finds the global (or patch-local) node numbers on a patch boundary.
+  //! \param[in] lIndex Local index of the boundary edge
+  //! \param nodes Array of node numbers
```
Comment (kmokstad) [inline src/ASM/LR/ASMu2Dmx.h]: Mark `nodes` as [out]

File: src/ASM/LR/ASMu3D.h (hunk 3 above)
Comment (kmokstad) [inline src/ASM/LR/ASMu3D.h]: And the output argument `nodes` as [out].

File: src/ASM/LR/ASMu3D.C [hunk 9]
```diff
@@ -1962,21 +1950,26 @@ std::vector<int> ASMu3D::getFaceNodes (int face, int basis, int orient) const
...
 
 #if SP_DEBUG > 1
   std::cout <<"Boundary nodes in patch "<< idx+1 <<" edge "<< lIndex <<":";
   for (size_t i = 0; i < nodes.size(); i++)
     std::cout <<" "<< nodes[i];
   std::cout << std::endl;
 #endif
+
+  if (!local)
+    for (auto& node : nodes)
```
Comment (kmokstad) [inline src/ASM/LR/ASMu3D.C]: `auto` is convenient, but when the actual type is simple, as in this case, I prefer to use that instead, for readability. I.e., in this case `for (int& node : nodes)`

File: src/ASM/ASMs3Dmx.h [hunk 10]
```diff
@@ -223,7 +223,7 @@ class ASMs3Dmx : public ASMs3D, private ASMmxBase
   //! \param thick Thickness of connection
   //! \param local If \e true return patch-local node numbers
   virtual void getBoundaryNodes(int lIndex, IntVec& nodes, int basis = 0,
```
Comment (kmokstad) [inline src/ASM/ASMs3Dmx.h]: Mark arguments as [in] and `nodes`as [out]

File: src/ASM/ASMs3D.h [hunk 11]
```diff
@@ -203,12 +203,12 @@ class ASMs3D : public ASMstruct, public ASM3D
   //! \param local If true return patch-local node numbers
```
Comment (kmokstad) [inline src/ASM/ASMs3D.h]: Mark arguments as [in] and nodes as [out]

File: src/ASM/ASMs2Dmx.h [hunk 12]
```diff
@@ -219,7 +219,7 @@ class ASMs2Dmx : public ASMs2D, private ASMmxBase
   //! \param thick Thickness of connection
   //! \param local If \e true return patch-local node numbers
```
Comment (kmokstad) [inline src/ASM/ASMs2Dmx.h]: Mark arguments as [in] and `nodes` as [out]

File: src/ASM/ASMs2D.h [hunk 13]
```diff
@@ -187,7 +187,7 @@ class ASMs2D : public ASMstruct, public ASM2D
   //! \param local If \e true return patch-local node numbers
```
Comment (kmokstad) [inline src/ASM/ASMs2D.h]: Mark arguments as [in] and `nodes` as [out]

File: src/ASM/ASMs1D.h [hunk 14]
```diff
@@ -111,7 +111,7 @@ class ASMs1D : public ASMstruct, public ASM1D
   //! \param thick Thickness of connection
   //! \param local If \e true, return patch-local node numbers
```
Comment (kmokstad) [inline src/ASM/ASMs1D.h]: Mark arguments as [in] and nodes as [out]

**Comment by kmokstad** [comment]
In my previous review, please disregard all "mark nodes as [out]" since it actually is both in and out. But all the other arguments are [in].

**Comment by akva2** [comment]
jenkins build this with downstreams ifem-advectiondiffusion=14 please

**Comment by akva2** [comment]
contents of the pr has been reduced to the original intention, reviewed and tests pass. merging.

**Pull request merged**